internals/overlord/logstate/loki/loki.go

```diff
@@ -169,9 +196,7 @@ func (c *Client) buildRequest() lokiRequest {
 	for _, service := range services {
 		entries := bucketedEntries[service]
 		stream := lokiStream{
-			Labels: map[string]string{
-				"pebble_service": service,
-			},
+			Labels:  c.getLabels(service),
```

Comment:
As we discussed in our call, which we should discuss next week: Since the labels are based on environment, and are lazily evaluated here + we reuse the ring buffer between service instances + we have a common shared buffer in the client, we have the risk that a services's logs emitted before a service restart will be labelled the same as logs emitted after the restart.

Reply:
The obvious/easy fix is to force flush the client whenever the plan changes.

However, there is one potential issue - the PlanChanged is called from a separate goroutine to the gatherer's main control loop. So there might be concurrency issues here (since we didn't design the client to be concurrency-safe).